MaxText/max_utils.py

```diff
@@ -360,6 +365,65 @@ def fill_unspecified_mesh_axes(parallelism_vals, target_product, parallelism_typ
   return parallelism_vals
 
 
+def create_custom_64x4_device_mesh(
+    mesh_shape: Sequence[int],
+    dcn_mesh_shape: Sequence[int],
+    devices: Sequence[Any],
+    process_is_granule: bool = False,
+    should_sort_granules_by_key: bool = True,
+) -> np.ndarray:
+  assert devices % 256 == 0, f"This custom mesh is not valid for {devices} devices"
+  attr = 'process_index' if process_is_granule else 'slice_index'
+  if not hasattr(devices[0], attr):
+    raise ValueError(
+        f'Device {devices[0]} does not have attribute {attr}. See'
+        ' `process_is_granule` option.'
+    )
+  granule_dict = collections.defaultdict(list)
+  for dev in devices:
+    granule_dict[getattr(dev, attr)].append(dev)
+  granules = (
+      [granule_dict[key] for key in sorted(granule_dict.keys())]
+      if should_sort_granules_by_key
+      else granule_dict.values()
+  )
+  if np.prod(dcn_mesh_shape) != len(granules):
+    raise ValueError(
+        f'Number of slices {len(granules)} must equal the product of '
+        f'dcn_mesh_shape {dcn_mesh_shape}'
+    )
+  per_granule_meshes = [
+      mesh_utils.create_device_mesh(
+          [16,16],
+          granule,
+          allow_split_physical_axes=False,
+      )
+      for granule in granules
+  ]
+  def reshape_mesh_to_rings(a):
+    b = []
+    for i in range(8):
+      b.append([])
+      for j in range(8):
+        a_i = i*2
+        a_j = j*2
+        # forms a ring of size 4
+        b[i].append([a[a_i,a_j], a[a_i,a_j+1], a[a_i+1,a_j+1], a[a_i+1,a_j]])
+    b = np.array(b)
+    b = np.reshape(b, (64,4))
```

Comment:
I wish we could apply this to other mesh shapes besides 64,4 easily but I don't know the underlying rules enough to make this modular.

Reply:
Ya I think we should chat with compiler folks to understand this ourselves better. Are we able to create an efficient 8x32 mesh as well?

Reply:
I saw Tynan's CL where he created this change. The logic is a little different, no rings are created for 32x8 mesh shape.

Reply:
Wondering did folks already file a feature request to JAX and/or if there's already work ongoing to generate more efficient meshes by default?